File: crud_java/src/main/java/com/example/crud_java/presentation/item/ItemController.java
package com.example.crud_java.presentation.item;

import com.example.crud_java.application.common.dto.PageDto;
import com.example.crud_java.application.item.ItemService;
import com.example.crud_java.application.item.dto.*;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.data.web.PageableDefault;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.net.URI;

@RestController
@RequiredArgsConstructor
@RequestMapping("/v1/items")
public class ItemController {
    private final ItemService itemService;

    @GetMapping
    public ResponseEntity<PageDto> readItems(
            @PageableDefault(size = 20) Pageable pageable,
            ItemReadRequest request
    ) {
        PageDto response = itemService.readItems(pageable, request);
        return ResponseEntity.ok(response);
    }

    @PostMapping
    public ResponseEntity<Void> createItem(@RequestBody ItemCreateRequest request) {
        Long itemNo = itemService.createItem(request);
        URI uri = URI.create(String.format("/v1/items/%d", itemNo));
        return ResponseEntity.created(uri).build();
    }

    @GetMapping("/{itemNo}")
    public ResponseEntity<ItemReadResponse> readItem(@PathVariable Long itemNo) {
        ItemReadResponse response = itemService.readItem(itemNo);
        return ResponseEntity.ok(response);
    }

    @PutMapping("/{itemNo}")
    public ResponseEntity<Void> updateItem(
            @PathVariable Long itemNo,
            @RequestBody ItemUpdateRequest request
    ) {
        itemService.updateItem(itemNo, request);
        return ResponseEntity.noContent().build();
    }

    @PatchMapping("/{itemNo}")
    public ResponseEntity<Void> updateItemQuantity(
            @PathVariable Long itemNo,
            @RequestBody ItemQuantityUpdateRequest request
    ) {
        itemService.updateItemQuantity(itemNo, request);
        return ResponseEntity.noContent().build();
    }

    @DeleteMapping("/{itemNo}")
    public ResponseEntity<Void> deleteItem(@PathVariable Long itemNo) {
        itemService.deleteItem(itemNo);
        return ResponseEntity.noContent().build();
    }
}

File: crud_kotlin/src/main/kotlin/com/example/crud_kotlin/presentation/item/ItemController.kt
package com.example.crud_kotlin.presentation.item

import com.example.crud_kotlin.application.common.dto.PageDto
import com.example.crud_kotlin.application.item.ItemService
import com.example.crud_kotlin.application.item.dto.*
import org.springframework.data.domain.Pageable
import org.springframework.data.web.PageableDefault
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.net.URI

@RestController
@RequestMapping("/v1/items")
class ItemController(
    private val itemService: ItemService
) {

    @GetMapping
    fun readItems(
        @PageableDefault(size = 20) pageable: Pageable,
        request: ItemReadRequest
    ): ResponseEntity<PageDto<Iterable<ItemReadResponse>>> =
        ResponseEntity.ok(
            itemService.readItems(pageable, request)
        )

    @PostMapping
    fun createItem(@RequestBody request: ItemCreateRequest): ResponseEntity<Unit> {
        val itemNo = itemService.createItem(request)
        val uri = URI("/v1/items/$itemNo")
        return ResponseEntity.created(uri).build()
    }

    @GetMapping("/{itemNo}")
    fun readItem(@PathVariable itemNo: Long): ResponseEntity<ItemReadResponse> =
        ResponseEntity.ok(
            itemService.readItem(itemNo)
        )

    @PutMapping("/{itemNo}")
    fun updateItem(
        @PathVariable itemNo: Long,
        @RequestBody request: ItemUpdateRequest
    ): ResponseEntity<Unit> {
        itemService.updateItem(itemNo, request)
        return ResponseEntity.noContent().build()
    }

    @PatchMapping("/{itemNo}")
    fun updateItemQuantity(
        @PathVariable itemNo: Long,
        @RequestBody request: ItemUpdateQuantityRequest
    ): ResponseEntity<Unit> {
        itemService.updateItemQuantity(itemNo, request)
        return ResponseEntity.noContent().build()
    }

    @DeleteMapping("/{itemNo}")
    fun deleteItem(@PathVariable itemNo: Long): ResponseEntity<Unit> {
        itemService.deleteItem(itemNo)
        return ResponseEntity.noContent().build()
    }
}